bson/src/main/org/bson/json/JsonObject.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.json;
+
+import org.bson.BsonDocument;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+
+/**
+ * A wrapper class that holds a JSON object string. This class makes decoding JSON efficient.
+ * Note that this class only holds valid JSON objects, not arrays.
+ *
+ * @since 4.2
+ */
+public class JsonObject implements Bson {
+    private final String json;
+
+    /**
+     * Constructs a new instance with the given JSON object string. Clients must ensure
+     * they only pass in valid JSON objects to this constructor. The constructor does not
+     * perform full validation on construction, but an invalid JsonObject can cause errors
+     * when it is used later on.
+     *
+     * @param json the JSON object string
+     */
+    public JsonObject(final String json) {
+        if (json == null) {
+            throw new IllegalArgumentException("Json cannot be null");
+        }
+
+        String trimmed = json.trim();
```

Comment:
While use of `trim()` is convenient, it results in a full copy of the string, which is unnecessarily wasteful.  I suggest instead to write a simple loop to skip over any leading whitespace characters defined in the JSON spec (20, A, D, 9) and ensure the first non-whitespace character is `{`.  I don't think it's worth ensuring that the string ends with `}`.  Since we're not checking the validity of the JSON here, it's not really necessary (if the string doesn't end with `}` it's not valid JSON anyway and will throw a parse exception eventually)

Reply:
Also, technically, the `trim()` method defines whitespace differently from the JSON spec.

Reply:
Done